Table: Inline select and date select use portals!

dancingshell opened the pull request:
### Summary
<!--Why you made the changes-->
<!--What is the intended result-->
Dropdowns for inline components were pretty broken in small table datasets or near the bottom of the table. This PR takes advantage of the new react portal feature in react 16
- Updated to React 16.x.x
- `InlineSelect` and `InlineDateSelect` now use react portals for better popout positioning
- rearranged functions a bunch in `InlineDateSelect` to make it easier to read
- tests now pass

### Related Issues
Closes: <!--Issues it closes if any-->  
Relates To: <!--Related Pull Requests/Issues-->

### Screenshots
#### Before
![screen shot 2017-11-07 at 4 42 23 pm](https://user-images.githubusercontent.com/6993945/32525735-bf7eafbc-c3da-11e7-8e14-47f1935fa714.png)

#### After
![screen shot 2017-11-07 at 4 42 00 pm](https://user-images.githubusercontent.com/6993945/32525745-c451f4cc-c3da-11e7-998a-01b91e865bde.png)

### Review Notes/ Questions/ TODOs
<!-- Call out anything you want the reviewer to be aware of -->
Since the portal content (dropdown content) is now no longer nested in the parent container class, some styles need to be re-implemented in the dropdown itself
### Testing
<!-- Steps to reproduce -->

mcclayton commented on table/README.md:
```diff
@@ -4,11 +4,16 @@ A table component with checkboxes, sticky head, sticky right panel, and horizont
 
 ##### 9.2.2
 - `stickyHeader` prop added
+>>>>>>> master
```
Looks like an artifact of a merge resolution?

ben-burlingham commented on table/src/InlineDateSelect.scss:
```diff
@@ -114,6 +114,11 @@
+
+  * {
+    box-sizing: border-box;
+    font: 11px sans-serif;
```
nb Might not want to put a font declaration in the universal selector; it's quite hard to override and will probably cause a mismatched font error in any production stack

dancingshell replied:
even when it is nested inside another class?

ben-burlingham replied:
That arguably makes it worse, unfortunately, because anyone who needs to override it will have to try harder.  Star selectors are generally evil for a variety of reasons - performance, specificity, and lack of precision.

ben-burlingham commented on table/src/InlinePortal.jsx:
```diff
@@ -0,0 +1,45 @@
+  componentWillUnmount() {
+    this.closePortal();
+  }
+  createPortal(props) {
```
nb linting?

ben-burlingham commented on table/src/Table.scss:
```diff
@@ -1,6 +1,6 @@
 @import '../../shared/variables';
 
-$headerZ: 2;
+$headerZ: 6;
```
nb Is this a carefully calculated value, and not a shotgun?

chrisarevir commented:
The only issue I see with this is that in order to take advantage of any of the new portal stuff is that the client needs to be on react16 for it to work. Particles do not ship with their own version of react so it might end up breaking/not working properly if a client is not "up to date". See the shared webpack.config.js at the root and notice the "external" key names react and react-dom.

chrisarevir commented on table/package.json:
```diff
@@ -1,13 +1,13 @@
 {
   "name": "@particles/table",
-  "version": "9.4.0",
+  "version": "9.4.1",
```
This would be a breaking change if now on react 16 since the API for the needed parts is no longer there.

chrisarevir commented on table/src/InlineDateSelect.jsx:
```diff
@@ -137,16 +139,15 @@ export default class InlineDateSelect extends React.Component {
     document.activeElement.blur();
     this.setState({ expanded: false });
-  }
+  };
```
Needed?

chrisarevir commented on table/src/InlineDateSelect.jsx:
```diff
@@ -296,7 +285,7 @@ export default class InlineDateSelect extends React.Component {
     this.setLeft(dropdownLeft, tableLeft);
     this.setTop(dropdownTop, tableBottom, tableTop);
-  }
+  };
```
Needed?

dancingshell commented:
Updated to major version bump. Also included:
- Date picker now opens to initial value if passed in via prop and highlights date
- Fixed bug where clicking inline date title didn't open picker
